### PR TITLE
Close actions & sources #2624

### PR DESCRIPTION
**Pull request by feyzullah**
Added new sources;
* new-lead
* lead-status-changed
* new-opportunity
* custom-event

Added new actions;
* search-leads
* create-lead
* update-lead
* custom-action

### CODE REVIEW

**Comment by dylburger** [comment]
Thanks for submitting this PR! When we review PRs, we follow the [Pipedream component guidelines](https://pipedream.com/docs/components/guidelines/). If you're not familiar, here's a quick checklist:

- [ ] [Create components to address specific use cases whenever possible](https://pipedream.com/docs/components/guidelines/#component-scope)
- [ ] [Component `key`s should follow the format `app_name_slug-slugified-component-name`](https://pipedream.com/docs/components/guidelines/#component-key-pattern)
- [ ] Components should follow the [standard directory structure](https://pipedream.com/docs/components/guidelines/#folder-structure)
- [ ] [Prefer Node.js client libraries to REST APIs](https://pipedream.com/docs/components/guidelines/#using-apis-vs-client-libraries)
- [ ] When making API requests, [handle pagination](https://pipedream.com/docs/components/guidelines/#pagination) to ensure all data / events are processed
- [ ] [Use `secret` props](https://pipedream.com/docs/components/guidelines/#capturing-sensitive-data) to capture sensitive data
- [ ] Props and methods should be defined in [app files](https://pipedream.com/docs/components/guidelines/#app-files) whenever possible
- [ ] Document methods with [JS Docs](https://pipedream.com/docs/components/guidelines/#methods)
- [ ] [Use `optional` props](https://pipedream.com/docs/components/guidelines/#optional-vs-required-props) whenever possible, and [set a `default` value](https://pipedream.com/docs/components/guidelines/#default-values) where you can
- [ ] Use [async options](https://pipedream.com/docs/components/guidelines/#async-options) to accept user input wherever possible
- [ ] [Create subscription webhooks sources (vs polling sources) whenever possible](https://pipedream.com/docs/components/guidelines/#webhook-vs-polling-sources)
- [ ] [Always emit a `summary` with events](https://pipedream.com/docs/components/guidelines/#emit-a-summary)
- [ ] [Use a built-in deduping strategy wherever possible](https://pipedream.com/docs/components/guidelines/#deduping)
- [ ] When building webhook sources, [follow these best practices](https://pipedream.com/docs/components/guidelines/#webhook-sources)
- [ ] When building polling sources, [follow these best practices](https://pipedream.com/docs/components/guidelines/#polling-sources)

File: components/close/close.app.mjs [hunk 1]
```diff
@@ -1,11 +1,141 @@
+import axios from "axios";
```
[the remark this answers is not in the document]
Comment (feyzullah) [reply components/close/close.app.mjs:1]: Hi @jcortes I cannot migrate to pipedream platform axios. Please see the convo [here](https://pipedream-users.slack.com/archives/C01E5KCTR16/p1651082655387479). For the API client, before moving forward I checked it and I think it is not easier to read than the axios solution. Please let me know what you think about this. For the others, I will be doing as you suggested, thanks!

File: components/close/close.app.mjs [hunk 2]
```diff
@@ -1,11 +1,141 @@
...
+      }
+      return parsed;
+    },
+    async createHook({ ...args } = {}) {
+      const response = await this._makeRequest({
+        method: "POST",
+        path: "/webhook",
+        ...args,
+      });
+      return response;
```
[the remark this answers is not in the document]
Comment (feyzullah) [reply components/close/close.app.mjs]: Done

File: components/close/close.app.mjs [hunk 3]
```diff
@@ -1,11 +1,141 @@
...
+    },
+    parseObject(obj) {
+      let parsed;
+      try {
+        parsed = JSON.parse(obj);
+      } catch (e) {
+        parsed = obj;
+      }
+      return parsed;
+    },
```
[the remark this answers is not in the document]
Comment (feyzullah) [reply components/close/close.app.mjs]: Done

File: components/close/close.app.mjs [hunk 4]
```diff
@@ -1,11 +1,141 @@
+import axios from "axios";
+
 export default {
   type: "app",
   app: "close",
-  propDefinitions: {},
+  propDefinitions: {
+    statusId: {
+      label: "Status ID",
+      description: "TODO",
```
[the remark this answers is not in the document]
Comment (feyzullah) [reply components/close/close.app.mjs]: Done

File: components/close/sources/common.mjs [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+  },
+  methods: {
+    async activate({ events }) {
+      const response = await this.close.createHook({
+        data: {
+          events,
+          url: this.http.endpoint,
+        },
+      });
+      this.db.set("hookId", response.data.id);
```
[the remark this answers is not in the document]
Comment (feyzullah) [reply components/close/sources/common.mjs]: Done

File: components/close/sources/custom-event/custom-event.mjs [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+    async activate() {
+      await this.activate({
+        events: [
+          {
+            object_type: this.objectType,
+            action: this.action,
+          },
+        ],
+      });
+    },
```
[the remark this answers is not in the document]
Comment (feyzullah) [reply components/close/sources/custom-event/custom-event.mjs]: Done

File: components/close/actions/create-lead/create-lead.mjs [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+      description: "Please provide an object array e.g., \n\
+        [\n { \n\
+          \"name\": \"Gob\",\n\
+          \"title\": \"Sr. Vice President\",\n\
+          \"emails\": [ { \"type\": \"office\", \"email\": \"gob@example.com\" } ],\n\
+          \"phones\": [ { \"type\": \"office\", \"phone\": \"8004445555\" } ] \n\
+        }\n]",
+      type: "string[]",
+      optional: true,
+    },
```
[the remark this answers is not in the document]
Comment (feyzullah) [reply components/close/actions/create-lead/create-lead.mjs]: Done

File: components/close/package.json [hunk 8]
```diff
@@ -0,0 +1,17 @@
+{
+  "name": "@pipedream/close",
+  "version": "0.0.1",
```
[the remark this answers is not in the document]
Comment (feyzullah) [reply components/close/package.json:3]: Done

File: components/close/close.app.mjs [hunk 9]
```diff
@@ -1,11 +1,141 @@
...
+    parseObject(obj) {
+      let parsed;
+      try {
+        parsed = JSON.parse(obj);
+      } catch (e) {
+        parsed = obj;
+      }
+      return parsed;
+    },
+    async createHook({ ...args } = {}) {
```
[the remark this answers is not in the document]
Comment (feyzullah) [reply components/close/close.app.mjs]: Done

File: components/close/close.app.mjs [hunk 10]
```diff
@@ -1,11 +1,131 @@
...
+      ...args
+    } = {}) {
+      return await this._makeRequest({
+        method: "PUT",
+        path: `/lead/${leadId}`,
+        ...args,
+      });
+    },
+    async searchLeads(args = {}) {
+      return await this._makeRequest({
```
Comment (jcortes) [inline components/close/close.app.mjs:124]:
```suggestion
      return this._makeRequest({
```

File: components/close/close.app.mjs [hunk 11]
```diff
@@ -1,11 +1,131 @@
...
+    },
+    async listLeadStatus(args = {}) {
+      return await this._makeRequest({
+        method: "GET",
+        path: "/status/lead",
+        ...args,
+      });
+    },
+    async createLead(args = {}) {
+      return await this._makeRequest({
```
Comment (jcortes) [inline components/close/close.app.mjs:107]:
```suggestion
      return this._makeRequest({
```

File: components/close/close.app.mjs [hunk 12]
```diff
@@ -1,11 +1,131 @@
...
+    },
+    async listLeads(args = {}) {
+      return await this._makeRequest({
+        method: "GET",
+        path: "/lead",
+        ...args,
+      });
+    },
+    async listLeadStatus(args = {}) {
+      return await this._makeRequest({
```
Comment (jcortes) [inline components/close/close.app.mjs:100]:
```suggestion
      return this._makeRequest({
```

File: components/close/close.app.mjs [hunk 13]
```diff
@@ -1,11 +1,131 @@
...
+      ...args
+    } = {}) {
+      return await this._makeRequest({
+        method: "DELETE",
+        path: `/webhook/${hookId}`,
+        ...args,
+      });
+    },
+    async listLeads(args = {}) {
+      return await this._makeRequest({
```
Comment (jcortes) [inline components/close/close.app.mjs:93]:
```suggestion
      return this._makeRequest({
```

File: components/close/close.app.mjs [hunk 14]
```diff
@@ -1,11 +1,131 @@
...
+      const config = {
+        url: this._getUrl(path),
+        headers: this._getHeaders(headers),
+        auth: this._getAuthKeys(),
+        ...otherConfig,
+      };
+      return axios(config);
+    },
+    async createHook(args = {}) {
+      return await this._makeRequest({
```
Comment (jcortes) [inline components/close/close.app.mjs:76]:
```suggestion
      return this._makeRequest({
```

**Review by jcortes — CHANGES_REQUESTED** [review]
Hi @feyzullah I've just added a few minor request changes and it will be ready for QA! Thanks for making all these changes!

File: components/close/close.app.mjs (hunk 14 above)
Comment (feyzullah) [reply components/close/close.app.mjs]: @jcortes they are really unnecessary. Sorry for that. I deleted them, thank you!

**Review by jcortes — DISMISSED** [review]
@feyzullah This PR is ready for QA. Thanks!

**Comment by DilanAthukorala** [comment]
@feyzullah Observing following error, could you please check if the contacts object is being passed correctly.

image

**Review by jcortes — APPROVED** [review]
LGTM!

**Comment by feyzullah** [comment]
> @feyzullah Observing following error, could you please check if the contacts object is being passed correctly.
> 
> image

Hi @DilanAthukorala I guess there was something broken after refactoring. It must have been fixed. Can you please check again?

**Comment by DilanAthukorala** [comment]
Verification completed. 

cc: @dannyroosevelt

**Pull request merged**